packages/truffle-codec-utils/src/definition.ts

```diff
@@ -403,8 +412,254 @@ export namespace Definition {
+    if(checkIndexed) {
+      indexed = node.indexed; //note: may be undefined for a base type
+    }
+    //is this an array? if so use separate logic
+    if(typeClass(node) === "array") {
+      let baseType = node.typeName ? node.typeName.baseType : node.baseType;
+      let baseAbi = parameterToAbi(baseType, referenceDeclarations, checkIndexed);
+      let arraySuffix = isDynamicArray(node)
+        ? `[]`
+        : `[${staticLength(node)}]`;
+      return {
+        name,
+        type: baseAbi.type + arraySuffix,
+        indexed,
+        components: baseAbi.components
+      };
+    }
+    let abiTypeString = toAbiType(node, referenceDeclarations);
+    //otherwise... is it a struct? if so we need to populate components
+    if(typeClass(node) === "struct") {
+      let id = typeId(node);
+      let referenceDeclaration = referenceDeclarations[id];
+      if(referenceDeclaration === undefined) {
+        let typeToDisplay = typeString(node);
+        throw new UnknownUserDefinedTypeError(id, typeToDisplay);
       }
+      components = parametersToAbi(referenceDeclaration.members, referenceDeclarations, checkIndexed);
+    }
+    return {
+      name, //may be empty string but should only be undefined in recursive calls
+      type: abiTypeString,
+      indexed, //undefined if !checkedIndex
+      components //undefined if not a struct or (multidim) array of structs
+    };
+  }
+
+  //note: this is only meant for non-array types that can go in the ABI
+  //it returns how that type is notated in the ABI -- just the string,
+  //to be clear, not components of tuples
+  //again, NOT FOR ARRAYS
+  function toAbiType(node: AstDefinition, referenceDeclarations: AstReferences): string {
+    let basicType = typeClassLongForm(node); //get that whole first segment!
+    switch(basicType) {
+      case "contract":
+        return "address";
+      case "struct":
+        return "tuple"; //the more detailed checking will be handled elsewhere
+      case "enum":
+        let referenceId = typeId(node);
+        let referenceDeclaration = referenceDeclarations[referenceId];
+        if(referenceDeclaration === undefined) {
+          let typeToDisplay = typeString(node);
+          throw new UnknownUserDefinedTypeError(referenceId, typeToDisplay);
+        }
+        let numOptions = referenceDeclaration.members.length;
+        let bits = 8 * Math.ceil(Math.log2(numOptions) / 8);
+        return `uint${bits}`;
+      default:
+        return basicType;
+        //note that: int/uint/fixed/ufixed/bytes will have their size and such left on;
+        //address will have "payable" left off;
+        //external functions will be reduced to "function" (and internal functions shouldn't
+        //be passed in!)
+        //(mappings shouldn't be passed in either obviously)
+        //(nor arrays :P )
+    }
+  }
+
+  function getterDefinitionToAbi(node: AstDefinition, referenceDeclarations: AstReferences): AbiUtils.FunctionAbiEntry {
```

Comment:
Hoo boy.  So, the `AstDefinition` type is currently a big mess.  I guess originally it was supposed to just represent definitions?  But we've been using it as just a general `AstNode` type.

So, it might make sense to redo all this to have different types of nodes, but, I wouldn't want to do that unless we're going to do that more thoroughly; I wouldn't want to *just* split off getters.

Note btw that getter definitions are just certain variable definitions; they're not some separate `nodeType`.  While obviously you could have one type when considering it as a variable and another when considering it as a getter, I dunno how well that works in TypeScript; I'm not sure it's really made for that sort of "these two things are different types because they mean different things even though their representation is the same" thing?  That may be me not knowing TypeScript sufficiently well, though.

Reply:
Sounds good to leave for now